### ISSUE TITLE
Idea: Dosbox and Debugger with ImGUI frontend

### ISSUE THREAD

**Issue by LowLevelMahn** [dosbox-staging/dosbox-staging#1045]
found this ImGUI based dosbox frontend

https://github.com/Patedam/DebugDosBox

a little outdated but still interesting

**Comment by kcgen** [comment]
Nice find! ImGUI is something @dreamer wants to incorporate for various GUI feedback and OSD purposes. 
This looks great.

I think the way to work on this would be to:
 1. I can create a feature branch (at the project level)
 1. Developers working on this would branch off this have write access to merge into it (plus the usual CI gates).
 1. The feature branch would be rebased against the master so it never falls behind or has risk of divergence. 
 1. Developers could work on this over many months, incrementally improving it.
 1. The initial goal would be to reach functional-parity with the existing debugger.
 1. Once it reaches parity and is stable, we'd merge the feature branch and drop the old debugger.

We've used this approach for SDL2, GUS, and Meson, as these were non-trivial changes that had to progress them over several months.

**Comment by kcgen** [comment]
References from @kklobe : 
 - https://sites.google.com/site/pcdosretro/enhdebug
 - https://www.scummvm.org/news/20180331/

**Comment by kcgen** [comment]
@GranMinigun reported that Ghidra has a debugger in current beta.

**Comment by LowLevelMahn** [comment]
the problem with real dos debuggers is that they mostly based on real hardware - for example SoftIce does not run under dosbox 

the best what i've seen so far is https://github.com/wjp/idados - a dosbox extension that allows debugging of 16/32bit software in IDA Pro using dosbox as the backend

**Comment by dreamer** [comment]
I do like this interface that @Patedam implemented  - yeah, our long-term goal is to get rid of debugger TUI and dependency on `curses` and have something very similar. Can't comment on code quality though, as the code was dumped in a big single commit in the repo.

**Comment by Patedam** [comment]
> 
> 
> I do like this interface that @Patedam implemented - yeah, our long-term goal is to get rid of debugger TUI and dependency on `curses` and have something very similar (but only for debugger builds - for normal builds I have something different in mind). Can't comment on code quality though, as the code was dumped in a big single commit in the repo.

Thanks.
I launched this project to debug something a while ago, code is very hackish but wanted to share the project in case someone wanted to use it. There are a lot of bugs too, but it was enough to do what i needed at the time :)

**Comment by dreamer** [comment]
@Patedam do you remember which SVN revision you used as a base for your implementation?

**Comment by LowLevelMahn** [comment]
could be tags/RELEASE_0_74_2 - revision 4156

**Comment by Patedam** [comment]
I'm not 100% sure but i have a zip of R4175 so that should be this one.

**Comment by LowLevelMahn** [comment]
did you also ported dosbox to SDL2 with these changes?

**Comment by Patedam** [comment]
Yes i did a very basic port to SDL2, i think i took a snippet of code on vogons.org but i dont remember which one.

**Comment by LowLevelMahn** [comment]
i would rebased into a dosbox staging svn/trunk 4175 based branch and
add a IMGUI_DEBUGGER #define that separates all the big and tiny changes from the base code - that does not work for the newly added cpp/hpp files but its then maybe easier to merge with more recent versions of svn/trunk
